### DISCUSSION TITLE
_middleware executes after getServerSideProps?

### DISCUSSION THREAD

**Discussion by akshay-nm** [vercel/next.js#37180] [Help] [answered]
I am trying to use _middleware as a basic authentication mech so that it adds a header to the request indicating whether it is authenticated or not. 
Trying to use it with API works, but trying to use it with getServerSideProps is not possible probably as my logs indicate that the _middleware is executed AFTER the getServerSideProps call. 

What is the execution order of a middleware, getServerSideProps and an API route call handler?

**Comment by icyJoseph** [comment, marked as answer]
~~I don't think there's a mechanism to share data from middleware to the application.~~

~~I mean that adding extra headers and such won't cut it. Unless you do redirects in the middleware.~~

~~Simply doing something like `req.headers.set('x-foo', 'bar')`, won't work. It is a bit of an issue I must say, but it is how it is.~~

A few things before answering the question at hand:

- We should only be doing root middleware, it is a beta feature, and in an upcoming release, nested middleware won't be supported
- Middleware catches all, even requests for `favicons` and stuff like that

So `middleware` runs before GSSP. Check in your logs `req.url`.

---

Curious about how are you sharing data between middleware and API routes?

## Edit

I've learned something today:

```ts
export const getServerSideProps = async (ctx: GetServerSidePropsContext) => {
  console.log(ctx.res.getHeader("x-browser-name"));
  return { props: {} };
};
```

And a middleware that does:

```ts
import { NextResponse, type NextRequest } from "next/server";

export function middleware(req: NextRequest) {
  const { pathname } = req.nextUrl;

  if (
    // pathname.startsWith("/api") || //  exclude all API routes
    pathname.startsWith("/static") || // exclude static files
    pathname.includes(".") // exclude all files in the public folder
  )
    return NextResponse.next();

  const userAgent = req.ua;

  if (!userAgent) return new Response(null, { status: 400 });

  const browserName = userAgent.browser.name;

  if (!browserName) return new Response(null, { status: 400 });

  const response = NextResponse.next();

  response.headers.set("x-browser-name", browserName);

  return response;
}
```

Ends up working out just fine. The `res` element in both API routes and GetServerSideProps, contains the header set by the middleware! 🎉 

Of course the entire browser name calculation is a bit of noise, but you see the point.

**Reply by akshay-nm** [reply]
I understand. 
I am using a root middleware. 
I refactored my logs and can now see that the `middleware` runs before everything.

Could you also shed some light on what happens when I return `NextResponse.next()` in the middleware? 
Is a new request initiated from the middleware to the backend?
In this case, is it possible to do something like this in the _middleware:
```js
const response = NextResponse.next()
// attach headers to response
return response
```
As this would mean, the backend would receive the information supplied via `res` headers as `req.headers`.

**Reply by icyJoseph** [reply]
I've updated my answer. 

I'd like to thank you, I setup a bit of time to investigate this further, and I got something to work.

If the middleware sets a header on `NextResponse.next()`, and returns that, then you can collect that header, not in the `req` parameter of API routes or GetServerSideProps, but on the response (`res`).

I'd still want to try it out deploying to Vercel, for instance, but this was great.

I'd be only concerned with changes to the middleware impacting this, perhaps this was unintended?

**Reply by akshay-nm** [reply]
Yes, attaching the values as custom headers (which being the app developer, you'd know which ones you can use safely), is a way you can share the information between `_middleware` and the backend.

**Reply by icyJoseph** [reply]
It's just counter intuitive that it is on the response object and not the request arriving at the server.

I'd say that's the end of this one :D fun experiment.

**Reply by akshay-nm** [reply]
Simple use-case/demo:
in root _middleware:
```js
import { NextResponse } from "next/server";

export default function middleware() {
  req.headers.set("x-sample-header", "some-sample-value");
  const response = NextResponse.next();
  response.headers.set("x-sample-header", "some-sample-value");
  return response;
}
```

then in some api/hello handler:
```js

export default function handler(req, res) {
  console.log(
    "Incoming headers from req: ",
    JSON.stringify(req.headers, null, 2)
  );
  console.log(
    "Outgoing headers from res: ",
    JSON.stringify(res.getHeaders(), null, 2)
  );
  res.removeHeader("x-sample-header");
  console.log(
    "After removing the headers: ",
    JSON.stringify(res.getHeaders(), null, 2)
  );

  res.status(200).json({ name: "John Doe" });
}
``` 

Then you get the attached header as part of `res` and not the one attached to `req`. The `res.removeHeader('x-sample-header')` also works.

**Reply by akshay-nm** [reply]
> It's just counter intuitive that it is on the response object and not the request arriving at the server.
> 
> I'd say that's the end of this one :D fun experiment.

Yes, there should be a more easy-to-use solution baked in nextjs if they are going forward with `_middleware`. 
Many of us who read middleware, think of something like express.js, where you can play with a `locals` if you need to share the data among different middleware.

I hope we can have something similar, if not better, by the time this is out of beta...
